php/php-rr-sdk.md

```diff
@@ -0,0 +1,692 @@
+{
+    yield new SignalHandler('name', function($input){
+        // do something
+    });       
+
+    yield new WaitSignal('name');
+}
+```
+
+### Sessions
+Sessions can be created as sub-activity group, using similar approach as `withTimeout` and other methods. However,
+the explicit yield required. 
+
+```php
+public function run(string $input): string
+{
+    $session = yield $this->activities->createSession(100, 100); // timeouts
+    
+    $result = yield $session->doSomething($input)->get();
+    $result = yield $session->somethingElse($result)->get();
+    
+    yield $session->complete();
+    
+    return $result;
+}
+```
+
+### Deterministic Time
+Workflows must avoid calling SPL functions `time()` and `date()`. Context method must be used instead:
+
+```php
+$ctx->getNow(); //DateTimeImmutable object.
+```
+
+### Side Effects
+Similar to Golang SDK the side effects can be registered using yield call:
+
+```php
+$result = yield new SideEffect(function(){
+    return mt_rand(0, 1000);
+});
+```
+
+Using syntax sugar (point of discussion):
+
+```php
+$result = yield $this->sideEffect(function(){
+    return mt_rand(0, 1000);
+});
+```
+
+> Similar to other SDKs, PHP version **does not** guarantee that sideEffect won't be called again during the replay. 
+
+### Error Handling
+Like in Java version of Temporal SDK the workflow errors can be delivered in form of exceptions triggered on `yield`:
+
+```php
+public function run(string $input): string
+{
+    try {
+        return yield $this->activities->doSomething($input)->get();
+    } catch (ActivityException $e) {
+        // get related activity ID
+        $this->logger->warning(
+            "Activity %s failed with %s",
+            $e->getActivityID(),
+            $e->getMessage()
+        );       
+    }
+}
+``` 
+
+Other errors can be triggered on workflow engine level, protocol error, etc. Exceptions must be diveded into following groups:
+
+Exception Type | Parent | Comment 
+--- | --- | ---
+EngineException | - | Low level exception, usually local.
+StalledException | EngineException | Triggered to indicate the workflow processing should stop (see below).
+WorkflowException | EngineException | Generic workflow error (like activity can not be created and etc)
+TimeoutException | WorkflowException | Workflow timed out (must contain information about actual timeout type). 
+TerminateException | WorkflowException | Workflow needs to be terminated.
+ActivityException | WorkflowException | Processing errors.
+ActivityTimeoutException | ActivityException | Activity processing timeouted (must contain information about actual timeout type). 
+
+> The exception tree is a subject of discussion. 
+
+### Termination and Cancel Requests
+The termination, cancel and stalled requested will be supplied to the workflow in a form of exceptions on `yield` call.
+Similar to activity errors workflow can handle these exceptions and perform a proper rollback.
```

Comment:
Termination is like "kill -9" it doesn't let workflow code to handle it.
Cancellation requires a separate discussion. What is the standard way in PHP to cancel things?

Reply:
There are not standard to cancel things, the process just dies. But since the workflow coordinator sits in another process we have to properly pass some signal to the underlying PHP process to clean up as well. 

TerminateException, in this case, allows us to gracefully stop the processing. It's not intended for handing, but rather to stop the coroutine without affecting other workflows in memory.

Reply:
I see. I would use a different name than Terminate just because Temporal service already uses Terminate for killing workflows.

Reply:
It can be variations of: 
PauseException
OffloadException

Since user don't need to handle it we can name a lot of freedom in naming it.